File: tests/zig/f2i.zig
const std = @import("std");
const profile = @import("profile.zig");

const total_numbers = 10000;
const total_times = 4;
var nums: [total_numbers][]const u8 = undefined;
var storage: [total_numbers][32]u8 = undefined;
var result: f32 = 0;
var times: u32 = total_times;

fn mkFloat(i: usize) f32 {
    const f: f32 = @floatFromInt(i);
    return f + f / @as(f32, total_numbers);
}

noinline fn f2i(rounds: u32) f32 {
    var sum: f32 = 0;
    for (0..rounds) |_| {
        for (nums) |s| {
            sum += std.fmt.parseFloat(f32, s) catch 0;
        }
    }
    return sum;
}

fn run(_: void) void {
    result = f2i(@as(*volatile u32, &times).*);
    std.mem.doNotOptimizeAway(result);
}

pub fn main(init: std.process.Init) !void {
    for (0..total_numbers) |i| {
        nums[i] = try std.fmt.bufPrint(&storage[i], "{d}", .{mkFloat(i)});
    }
    var reference: f32 = 0;
    for (0..total_times) |_| {
        for (0..total_numbers) |i| reference += mkFloat(i);
    }
    try profile.profile(init.io, "string2float", run, {});
    if (result != reference) profile.fail("string2float: sum differs from the reference");
}

File: tests/cpp/f2i.cpp
#include "profile.h"
#include <cmath>
#include <cstring>
#include <string>
#include <vector>

static const int TOTAL_NUMBERS = 10000;
static volatile int times = 4;

static float mk_float(int i) { return float(i) + float(i) / float(TOTAL_NUMBERS); }

PROFILE_NOINLINE float test_f2i(const std::vector<std::string> & nums, int rounds) {
    float summ = 0.0f;
    for (int i = 0; i != rounds; ++i) {
        for (const auto & s : nums) summ += float(atof(s.c_str()));
    }
    return summ;
}

int main() {
    std::vector<std::string> nums(TOTAL_NUMBERS);
    char buf[64];
    for (int i = 0; i < TOTAL_NUMBERS; i++) {
        snprintf(buf, sizeof(buf), "%.9g", mk_float(i));
        nums[i] = buf;
    }
    float reference = 0.0f;
    for (int j = 0; j < 4; j++) {
        for (int i = 0; i < TOTAL_NUMBERS; i++) reference += mk_float(i);
    }
    float result = 0.0f;
    profile::run("string2float", [&] { result = test_f2i(nums, times); profile::keep(result); });
    if (result != reference) profile::fail("string2float: sum differs from the reference");
    return 0;
}
